ci/cached-builds/gha_pr_changed_files.py

```diff
@@ -0,0 +1,115 @@
+import json
+import logging
+import os
+import pathlib
+import re
+import subprocess
+import unittest
+import urllib.request
+
+PROJECT_ROOT = pathlib.Path(__file__).parent.parent.parent.resolve()
+
+
+def get_github_token() -> str:
+    github_token = os.environ['GITHUB_TOKEN']
+    return github_token
+
+
+# https://docs.github.com/en/graphql/guides/forming-calls-with-graphql
+def compose_gh_api_request(pull_number: int, owner="opendatahub-io", repo="notebooks", per_page=100, cursor="") -> urllib.request.Request:
+    github_token = get_github_token()
+
+    return urllib.request.Request(
+        url="https://api.github.com/graphql",
+        method="POST",
+        headers={
+            "Authorization": f"bearer {github_token}",
+        },
+        # https://docs.github.com/en/graphql/guides/using-the-explorer
+        data=json.dumps({"query": f"""
+{{
+  repository(owner:"{owner}", name:"{repo}") {{
+    pullRequest(number:{pull_number}) {{
+      files(first:{per_page}, after:"{cursor}") {{
+        edges {{
+          node {{
+            path
+          }}
+          cursor
+        }}
+      }}
+    }}
+  }}
+}}
+    """}).encode("utf-8"),
+    )
+
+
+def list_changed_files(owner: str, repo: str, pr_number: int, per_page=100) -> list[str]:
+    files = []
+
+    logging.debug("Getting list of changed files from GitHub API")
+
+    CURSOR = ""
+    while CURSOR is not None:
+        request = compose_gh_api_request(pull_number=pr_number, owner=owner, repo=repo, per_page=per_page, cursor=CURSOR)
+        response = urllib.request.urlopen(request)
+        data = json.loads(response.read().decode("utf-8"))
+        response.close()
+        edges = data["data"]["repository"]["pullRequest"]["files"]["edges"]
+
+        CURSOR = None
+        for edge in edges:
+            files.append(edge["node"]["path"])
+            CURSOR = edge["cursor"]
```

Comment:
I think it would be more efficient to get the changed files using `git diff`.

Reply:
maybe? golangci action queries github api for it, see their docs for `only-new-issues`, https://github.com/golangci/golangci-lint-action?tab=readme-ov-file#only-new-issues

the gha query means I can fetch only the latest revision when pulling from git; and I don't have to worry about figuring out what's the PR base commit (I'd need to ask GH for that anyways, although that's available in the `github`  context right away https://docs.github.com/en/actions/learn-github-actions/contexts#github-context)